charms/knative-operator/src/charm.py

```diff
@@ -0,0 +1,134 @@
+                        "CONFIG_LOGGING_NAME": "config-loggig",
+                        "CONFIG_OBSERVABILITY_NAME": "config-observability",
+                    },
+                }
+            },
+        }
+        return Layer(layer_config)
+
+    def _update_layer(self, event) -> None:
+        """Updates the Pebble configuration layer if changed."""
+        if not self._container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for pod startup to complete")
+            event.defer()
+            return
+
+        # Get current config
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._knative_operator_layer
+        if current_layer.services != new_layer.services:
+            self._container.add_layer(self._operator_service, new_layer, combine=True)
+            logging.info("Pebble plan updated with new configuration")
+            self._container.restart(self._operator_service)
+
+    def _apply_all_resources(self) -> None:
+        """Helper method to create Kubernetes resources."""
+        for filename in self._resources_files:
+            manifest = self.env.get_template(filename).render(self._context)
+            for obj in codecs.load_all_yaml(manifest):
+                self.lightkube_client.apply(obj)
+
+    def _main(self, event):
+        """Event handler for changing Pebble configuration and applying k8s resources."""
+        try:
+            # Update Pebble configuration layer if it has changed
+            self.unit.status = MaintenanceStatus("Configuring Pebble layer")
+            self._update_layer(event)
+            self.unit.status = MaintenanceStatus("Applying resources")
+            # Apply Kubernetes resources
+            self._apply_all_resources()
+        except (pebble_APIError, lightkube_ApiError) as e:
+            if isinstance(e, pebble_APIError):
+                logging.error(traceback.format_exc())
+                self.unit.status = BlockedStatus("Error updating Pebble configuration layer")
+            elif isinstance(e, lightkube_ApiError):
+                logging.error(traceback.format_exc())
+                self.unit.status = BlockedStatus(
+                    f"Applying resources failed with code {str(e.status.code)}."
+                )
+                if e.status.code == 403:
+                    logging.error(
+                        "Received Forbidden (403) error when creating auth resources."
+                        "This may be due to the charm lacking permissions to create"
+                        "cluster-scoped resources."
+                        "Charm must be deployed with --trust"
+                    )
+                    event.defer()
+                    return
+        else:
+            self.unit.status = ActiveStatus()
+
+    def _on_knative_operator_pebble_ready(self, event):
+        """Event handler for on PebbleReadyEvent"""
+        self._update_layer(event)
```

Comment:
This has the same potential issue as the previous comment about setting `ActiveStatus`, but it feels like a more general possible issue in `_update_layer()` so I'm adding more comments there.

Reply:
@ca-scribner probably moving [this](https://github.com/canonical/knative-operators/pull/10/files#diff-e2c9ff5a819d559524f6f6fdbe84aa52d8b22eeb9f826f14e12c6455affef2ffR108) exception handling to `_update_layer()` helps.